Setting for Autorec start window behavior & settings rework

Glenn-1990 opened the pull request:
1) Added setting for the autorec start window behavior:
I disabled end time for when this setting is enabled, strictly speaking autorecs don't have an end time in tvh.
2) Reworked settings as there where lot's of categories and user could input invalid integers.

![schermafdruk van 2015-09-16 20 56 49](https://cloud.githubusercontent.com/assets/2036512/9915146/b23c79b2-5cb5-11e5-845a-7793664c5319.png)
![schermafdruk van 2015-09-16 20 56 22](https://cloud.githubusercontent.com/assets/2036512/9915147/b246884e-5cb5-11e5-8d57-eb99e3cb9b85.png)


@PhobosK, @Jalle19, @ksooo

Jalle19 commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -50,7 +50,25 @@ msgctxt "#30007"
 
-#empty strings from id 30008 to 30099
+msgctxt "#30008"
+msgid "Advanced Settings"
```
I'd write "settings" with a lower case "s".

Jalle19 replied:
Wait, never mind, it seems we've used this scheme previously too (e.g. "Data Transfer").

ksooo replied:
Nevertheless, I opt for lowercase. This is how we do in Kodi core.

ksooo commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -50,7 +50,25 @@ msgctxt "#30007"
+#empty strings from id 30009 to 30049
+
+msgctxt "#30050"
+msgid "Auto Recordings"
```
lowercase "r", please.

ksooo commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -50,7 +50,25 @@ msgctxt "#30007"
+msgstr ""
+
+msgctxt "#30051"
+msgid "Use timer start time as approximate"
```
I'd use tvheadend webui terminology here. Thus, it should be more understandable to users familiar with tvheadend webui.

Glenn-1990 replied:
Any suggestion? It's hard to get a good description in such a small string :-)

ksooo commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -153,7 +171,7 @@ msgstr ""
 
 msgctxt "#30400"
-msgid "Predictive tuning"
+msgid "Predictive Tuning"
```
lowercase "t", please

ksooo commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -161,5 +179,5 @@ msgid "Number of subscriptions to use"
 
 msgctxt "#30402"
-msgid "Unused subscription close delay"
+msgid "Unused subscription close delay (sec)"
```
Not sure about "sec" vs. "secs" here

ksooo commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -50,7 +50,25 @@ msgctxt "#30007"
+msgstr ""
+
+msgctxt "#30052"
+msgid "Max difference between approximate and real start (min)"
```
"min" vs "mins"?

ksooo commented:
For the records: The new feature has been discussed here: https://github.com/Glenn-1990/pvr.hts/commit/d71e0dd66e4d827b5b14846b8c8fa4490c37df66#commitcomment-13085861

ksooo commented on pvr.hts/resources/settings.xml:
```diff
@@ -10,16 +10,20 @@
-  <category label="30100">
+  <category label="30008">
+    <setting label="30050" type="lsep"/>
+	<setting id="autorec_approxtime" type="bool" label="30051" default="true"/>
```
Haha. default="false", please. PhobosK and I are in favour of the current approach, you are in favour of the new approach. 2 : 1 for the current approach. ;-)

Glenn-1990 replied:
If we get many complaints with "false" we could change it back to "true" right? :p

ksooo replied:
Sure.

ksooo commented on src/AutoRecordings.cpp:
```diff
@@ -191,23 +191,47 @@ PVR_ERROR AutoRecordings::SendAutorecAdd(const PVR_TIMER &timer)
+      /* Exact start time (minutes from midnight). */
+      struct tm *tm_start = localtime(&timer.startTime);
+      htsmsg_add_s32(m, "start", tm_start->tm_hour * 60 + tm_start->tm_min);
+    }
```
Isn't here

<pre>
else
    htsmsg_add_s32(m, "start", 25 * 60); // -1 or not sending causes server to set start and startWindow to any time
</pre>
missing?

Glenn-1990 replied:
/* Not sending causes server to set start and startWindow to any time */

And where does (25 * 60) come from? it seems magical :p

ksooo replied:
Magical or not, you're not allowed to change behavior. ;-)

ksooo commented on src/AutoRecordings.cpp:
```diff
@@ -191,23 +191,47 @@ PVR_ERROR AutoRecordings::SendAutorecAdd(const PVR_TIMER &timer)
+      /* Exact stop time (minutes from midnight). */
+      struct tm *tm_stop = localtime(&timer.endTime);
+      htsmsg_add_s32(m, "startWindow", tm_stop->tm_hour * 60 + tm_stop->tm_min);
+    }
```
Isn't here
<pre>
  else
    htsmsg_add_s32(m, "startWindow", 25 * 60); // -1 or not sending causes server to set start and startWindow to any time
</pre>
missing?

ksooo commented on src/HTSPTypes.cpp:
```diff
@@ -372,28 +372,58 @@ bool AutoRecording::operator!=(const AutoRecording &right)
+    else if (m_startWindowEnd < m_startWindowBegin)
+    {
+      /* End of start window is a day in the future */
+      int32_t newEnd  = m_startWindowEnd + ( 24 * 60 );
```
Remove the spaces between braces and numbers, please.

ksooo commented on src/HTSPTypes.cpp:
```diff
@@ -372,28 +372,58 @@ bool AutoRecording::operator!=(const AutoRecording &right)
-  if (m_startWindow == int32_t(-1)) // "any time"
+  if (tvh->GetSettings().bAutorecApproxTime)
+  {
+    /* Tvh doen't have an approximate stop time => "any time" */
```
doen't => doesn't

ksooo commented on src/client.cpp:
```diff
@@ -57,6 +57,8 @@ bool       g_bTraceDebug         = false;
 bool       g_bAsyncEpg           = false;
 int        g_iTotalTuners        = DEFAULT_TOTAL_TUNERS;
 int        g_iPreTunerCloseDelay = DEFAULT_PRETUNER_CLOSEDELAY;
+bool       g_bAutorecApproxTime  = true;
```
true => false

ksooo commented on src/client.cpp:
```diff
@@ -104,6 +106,10 @@ void ADDON_ReadSettings(void)
   UPDATE_INT(g_bTraceDebug, "trace_debug", false);
 
+  /* Auto recordings */
+  UPDATE_INT(g_bAutorecApproxTime, "autorec_approxtime", true);
```
true => false

ksooo commented:
Despite from my comments, looking good already.

I especially like the settings page redesign. Much appreciated. Thanks.

ksooo commented on pvr.hts/resources/settings.xml:
```diff
@@ -10,16 +10,20 @@
+	<setting id="autorec_maxdiff" enable="eq(-1,true)"  type="slider" option="int" range="0,5,120" label="30052" default="15" />
+  
+    <setting label="30400" type="lsep"/>
+	<setting id="total_tuners" type="enum" label="30401" values="1|2|3|4|5|6|7|8|9|10" default="0" />
```
I have 16 tuners! What now? Am I lost? ;-)

Enum seems not to be a good idea. As we are not able to obtain the number of tuners actually available, we don't know for sure which numbers are actually valid. So, we are doomed to the int we currently have for this, I'm afraid.

Glenn-1990 replied:
I just want to prevent that the user can input wrong (like 0) or unreal numbers (like 995442).

Actually the number of tuners is not elated to the number of subscriptions. (1 tuner can have multiple subscriptions). 

I just want to make this setting somehow fail save, 
i.e. if the user enters 1000 for example, everything seems fine to him until he starts zapping, the backend will be stressed and the network will be overloaded causing the tv stream to stall.
And the user might create a bug report for this...

10 subscriptions of +- 10Mbit per channel = 100Mbit.
Assuming that most users have at least a Mbit connection, this should be fail safe..

ksooo replied:
Not sure I like this, but get your point. No objections from my side. ... and I was just kidding about the 16 tuners I have. ;-)

ksooo commented on pvr.hts/resources/settings.xml:
```diff
@@ -10,16 +10,20 @@
+  
+    <setting label="30400" type="lsep"/>
+	<setting id="total_tuners" type="enum" label="30401" values="1|2|3|4|5|6|7|8|9|10" default="0" />
+	<setting id="pretuner_closedelay" enable="!eq(-1,0)" type="slider" option="int" range="5,5,60" label="30402" default="10" />
```
Where does the lower value of 5 for the range come from? Why not 0? Don't know, whether 0 makes actually sense. Just asking.

Glenn-1990 replied:
0 makes no sense as we close the old subscription immediately after a channel switch, we need at least a few seconds or predictive tuning won't work.

5 -> user must switch channels within 5 seconds to keep predictive tuning active, larger is better because is allows the user to switch channels more slowly. But 5 seconds as lower limit should be fine.

ksooo replied:
Okay. Thanks for clarifying.

Glenn-1990 commented:
@ksooo @Jalle19  Addressed all comments ;-)

ksooo commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -50,10 +50,28 @@ msgctxt "#30007"
+msgstr ""
+
+msgctxt "#30051"
+msgid "Use approximate start time instead of fixed start time"
```
This does not hit the point. Sounds, like the current approach would only match broadcasts starting at exactly at the given start time, which is not the case. I'm still squeezing my head to get something (imo) better out of it... ;-)

Glenn-1990 replied:
Yea, I know the feeling :p 
@Jalle19 do you have a suggestion?

PhobosK replied:
How about using a drop down menu with two choices named "Use relaxed start/end timings" / "Use strict start/end timings" that defaults to the first one which is the current implementation?
I think this would be proper and understandable for the user?
What do you think guys?

ksooo replied:
This is better, imo.

However, for the seeing as a list of choices we need a name for the seeing as well as names for the two values.

How about:

"Start time window calculation" for the setting's name and "Start + End time" (strict approach) and "Start +- < the name of the other new setting we should imo rename too >" (relaxed approach).

Should we rename the other new setting to "Fixed start time window span (+- mins)"?

Glenn-1990 replied:
If we have only 2 options, it may be better to use a spinner instead of a separate window?

ksooo replied:
The UI experts in the Kodi team say spinners are evil and should be avoided if not replaced everywhere. Not sure whether this holds for only 2 choices. @Phil65 any opinions on that?

Glenn-1990 replied:
If that is the case, I could also change the timeout settings (which are spinners) to get it all right in 1 PR?

ksooo replied:
Yep. Maybe in a separate commit.

phil65 replied:
Yep, gettin rid of spinners is much appreciated.  There are quite some more in PVR-related dialogs which can get replaced with better solutions. (timer and search dialog for example if I´m not mistaken)
In general the replacement options are:
- radiobuttons (for 2-state)
- DialogSelect
- button controls with a toggling label2

ksooo commented:
We need to find clever (and short) phrasing for the new approxtime setting. Despite from this, I'm fine with this PR now. Many thanks, @Glenn-1990

Glenn-1990 commented:
@ksooo What do you think of a Boolean setting to enable predictive tuning?
Now it looks like it's enabled with 1 subscription, which means that it's disabled..

Glenn-1990 commented:
@ksooo 
- added bool for predictive tuning so it should be more clear to the user
- added spinner and new names for the autorec settings
(a spinner or radio button is a must for the visibility control here)
- replaced some spinners with a slider

phil65 commented:
Would it be possible to exchange some of the sliders with a button which opens DialogNumeric? Slider controls in its current form is also something we want to get rid off for most parts.

Glenn-1990 commented:
@phil65 But what if we only want inputs between 1-60 for example?
The intention in this PR is to don't allow out the range inputs.

The choices for addon settings seems rather limited..

ksooo commented:
@Glenn-1990 I like all the changes as they are now. Thanks. Let's move the absolutely valid "sliders are evil, but there is no appropriate replacement available yet" debate to another PR, please. Don't want to make this one bigger and bigger (and therefore lasting longer and longer).

phil65 commented:
>and therefore lasting longer and longer

This PR is up for 3 days :smile: 

I only asked a question, there is no need to rush anything.

ksooo commented:
@phil65 

> This PR is up for 3 days :smile:

 Yeah. For pvr.hts we have another definition of "slow" than Kodi core has. :-)

> I only asked a question, there is no need to rush anything.

 All good. Thanks for your support and guidance.

phil65 commented:
All fine.
Back to topic: In case the fixed range is important for those settings then DialogNumeric indeed wont work out well. Leave those as sliders then for now.

Jalle19 commented on pvr.hts/resources/language/resource.language.en_gb/strings.po:
```diff
@@ -161,5 +187,9 @@ msgid "Number of subscriptions to use"
 
 msgctxt "#30402"
-msgid "Unused subscription close delay"
+msgid "Unused subscription close delay (secs)"
```
Sorry to continue the nitpicking but we should be consistent in how we indicate the unit of measurement. On the connection tab we use `Connection timeout in seconds`, I think we should either change that (and any others) to `Connection timeout (secs)` or change this one. Furthermore I like `(seconds)` better than `(secs)`, but it's not a big deal if it would make the setting name too long.

Jalle19 commented:
@Glenn-1990 can you address my last comment so we can get this in?

ksooo commented:
@Glenn-1990 and please add a changelog entry for existing version 2.2.8 (not a new version, please), thx. If this is done, you have my +1 for merging this.

Glenn-1990 commented:
@Jalle19 @ksooo done ;-)
